server/src/test/java/org/elasticsearch/cluster/routing/AllocationIdIT.java

```diff
@@ -0,0 +1,284 @@
+import java.util.Collection;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.index.query.QueryBuilders.matchAllQuery;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertHitCount;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.everyItem;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.isIn;
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.notNullValue;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.SUITE, numDataNodes = 0)
+public class AllocationIdIT extends ESIntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return Arrays.asList(MockTransportService.TestPlugin.class, MockEngineFactoryPlugin.class, InternalSettingsPlugin.class);
+    }
+
+    public void testFailedRecoveryOnAllocateStalePrimaryRequiresAnotherAllocateStalePrimary() throws Exception {
+        /* test case to spot the problem if allocation id is adjusted before recovery is done (historyUUID is adjusted):
+         * ( if recovery is failed on AllocateStalePrimary it requires another AllocateStalePrimary )
+         *
+         * - Master node + 2 data nodes
+         * - (1) index some docs
+         * - stop primary (node1)
+         * - (2) index more docs to a node2, that marks node1 as stale
+         * - stop node2
+         *
+         * node1 would not be a new primary due to master node state - it is required to run AllocateStalePrimary
+         * - put a corruption marker to node1 to interrupt recovery
+         * - start node1 (shard would not start as it is stale)
+         * - allocate stale primary - allocation id is adjusted (*) - but it fails due to the presence of corruption marker
+         * - stop node1
+         * - remove a corruption marker
+         * - try to open index on node1
+         *  -> node1 has a RED index if (*) points to a fake shard -> node requires another AllocateStalePrimary
+         * - check that restart does not help
+         * - another manual intervention of AllocateStalePrimary brings index to yellow state
+         * - bring node2 back
+         * -> nodes are fully in-sync and has different to initial history uuid
+         */
+
+        // initial set up
+        final String indexName = "index42";
+        final String master = internalCluster().startMasterOnlyNode();
+        String node1 = internalCluster().startNode();
+        createIndex(indexName);
+        final int numDocs = indexDocs(indexName, "foo", "bar");
+        final IndexSettings indexSettings = getIndexSettings(indexName, node1);
+        final String primaryNodeId = getNodeIdByName(node1);
```

Comment:
allocation commands can take node names, so I don't think we need this?

Reply:
oh, I did not know that, addressed in e8e3925